design-proposals/migration-policies.md

```diff
@@ -0,0 +1,224 @@
+* By “different groups of VMs” I could mean:
+    * All VMs that have certain label defined
+    * All VMs that have certain label with certain value defined
+    * All VMs that belong to a namespace with certain label key/value
+* As an admin I know that some workloads are in higher priority than others and therefore need different 
+  migration policies.
+* As an admin I know that some workloads are not migration-friendly and demand careful configurations in order
+  for migrations to converge.
+* As an admin I want to have the ability to temporarily tune migrations for a specific event such as draining
+  a node or performing an upgrade.
+* As an admin I want that the migration policy would be as transparent as possible for the VM owner / creator.
+* As a user I want to be able to tell admins the workload characteristics of my VMs so that admins can create
+  policies and/or namespaces with the desired characteristics.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+The design will be similar to Kubernetes' [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+which is a similar concept. Therefore, a new MigrationPolicy CRD will be introduced. This CRD's spec will contain all
+relevant fields from KubevirtCR's [MigrationConfiguration](https://kubevirt.io/api-reference/master/definitions.html#_v1_migrationconfiguration)
+(i.e. which are not inherently cluster-wide configurations). In the future, additional
+configurations (or status related fields) that inherently relate to non-global context could be added,
+thus the configurations on NetworkPolicy and KubevirtCR's MigrationConfigurations may diverge to some degree.
+
+In addition, the new CRD's spec will include the following ways of specifying the groups of VMIs on which
+to apply the policy:
+* By VMI's labels
+* By namespace's labels
+
+All of these methods can be combined, for example a policy can require both VMI labels and namespace labels in
+order to match.
+
+## Policies' Precedence
+
+It is possible that a multiple policies apply to the same VMI. In such cases, the precedence is in the
+same order as the bullets above (VMI labels first, then namespace labels). It is not allowed to define
+two policies with the exact same selectors.
+
+If multiple policies apply to the same VMI:
+* The most detailed policy will be applied, that is, the policy with the highest number of matching labels
+* If multiple policies match to a VMI with the same number of matching labels, the policies will be sorted by the 
+  lexicographic order of the matching labels keys. The first one in this order will be applied.
+  
+### Example
+
+For example, let's imagine a VMI with the following labels:
+* size: small
+* os: fedora
+* gpu: nvidia
+
+And let's say the namespace to which the VMI belongs contains the following labels:
+* priority: high
+* bandwidth: medium
+* hpc-workload: true
+
+The following policies are listed by their precedence (high to low):
```

Comment:
I went over https://github.com/kubevirt/kubevirt/pull/7322 which led me back to this here.

I don't understand why we went down this route. Even when reading the example it already overwhelms me. Further it basically cuts off all more advanced matchers (like not having a label). Why does this have to be different than all other selector mechanisms in k8s?

@davidvossel @iholder-redhat @vladikr 

I personally would drop this part and just match as always. If we want kind of a predictability on overlaps, I would go with name or something, but I would recommend to treat overlaps as errors from a user perspective.

Reply:
We had an offline discussion.

Looking at the above examples, to me, these don't reflect the desired direction of this feature. The original idea was to provide a way to apply a ceratin migration policy to a specific VMI or a group of VMIs that is "special"; in a way. In the first stage, we would need a user/admin cooperation to identify which VMIs are "special";
At the current state, we can rely on a provided label and a separate namespace where a prioritized group of VMIs wound live.

In the future, we would be able to target VMIs by different traits, such as their dirty rate, CPU usage, etc..., that would hint at the likelihood of successful migration and apply various methods to increase these chances. 

Letting the users influence the migration policy selection by applying a number of arbitrary labels would send a wrong message regarding how this feature should be used.